crates/sui-rosetta/src/types.rs

```diff
@@ -947,6 +959,63 @@ impl InternalOperation {
                 builder.pay_sui(recipients, amounts)?;
                 builder.finish()
             }
+            Self::PayCoin {
+                recipients,
+                amounts,
+                ..
+            } => {
+                let mut builder = ProgrammableTransactionBuilder::new();
+
+                let mut coins = metadata.objects.clone().into_iter();
+                let Some(coin) = coins.next() else {
+                    return Err(Error::InvalidInput("coins vector is empty".to_string()));
+                };
+                let coin_arg = builder.obj(ObjectArg::ImmOrOwnedObject(coin))?;
+                let merge_args: Vec<_> = coins
+                    .map(|c| builder.obj(ObjectArg::ImmOrOwnedObject(c)))
+                    .collect::<Result<_, _>>()?;
+                if !merge_args.is_empty() {
+                    builder.command(Command::MergeCoins(coin_arg, merge_args));
+                }
+                if recipients.len() != amounts.len() {
+                    return Err(Error::InvalidInput(
+                "Recipients and amounts mismatch. Got {recipients.len()} recipients but {amounts.len()} amounts".to_string()));
+                }
+
+                // collect recipients in the case where they are non-unique in order
+                // to minimize the number of transfers that must be performed
+                let mut recipient_map: IndexMap<SuiAddress, Vec<usize>> = IndexMap::new();
+                let mut amt_args = Vec::with_capacity(recipients.len());
+                for (i, (recipient, amount)) in recipients.into_iter().zip(amounts).enumerate() {
+                    recipient_map.entry(recipient).or_default().push(i);
+                    amt_args.push(builder.pure(amount)?);
+                }
+                let Argument::Result(split_primary) =
+                    builder.command(Command::SplitCoins(coin_arg, amt_args))
+                else {
+                    panic!("self.command should always give a Argument::Result")
+                };
+                let type_arg = TypeTag::from_str(&metadata.coin_type.clone().unwrap()).unwrap();
```

Comment:
isn't `&metadata.coin_type` already a TypeTag?

Reply:
nvm it isn't, it's a `StructTag`, but I think you can easily use `.into()`:
https://github.com/MystenLabs/sui/blob/32347c20c248f6691c6ab89ba53266e2f677ed9b/external-crates/move/crates/move-core-types/src/language_storage.rs#L368